### ISSUE TITLE
Streaming should not do "view building" on streamed base-table data 

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#11820]
This issue was discovered by @vladzcloudius and originally written in #7711, but that was a unrelated bug so I'm moving the new issue here.

The issue is about streaming (e.g., when a new node is added to the cluster) when tables have materialized views - doing an extra "view building step" which they shouldn't do, which slows streaming considerably.

As we know, in **repair**, we only repair base tables. So whatever base-table updates a base-replica receives during repair, it later also need to do "view building" (view updates) on these base writes, to update its paired view replica(s) as well.

But for **streaming** we are supposed do something very different:
1. We stream both base tables and view tables. 
2. When streaming base-table data, **no** "view building" or "view updates" should happen on this data.

Point 1 is criticial for correctness - often the base-table and view partition key are different, and the new node receives completely unrelated pieces of the base and view data - so it must stream the view tables separately and cannot in any way expect to deduce the view data from the base data it receive.

Point 2 is then only an issue of performance - since the new node already receives all the base and view data it is supposed to have, there is no point for it to do any view building.

If I understand @vladzcloudius's report correctly, he noticed that our current implementation does point 1 (so it's not in error) but fails on point 2 - it does unnecessary view building.

Is this actually true? Is this a regression (because I was sure it didn't work like this in the past)?

CC @asias @vladzcloudius

**Comment by vladzcloudius** [comment]
> @asias can you please take another look at the code and verify that streaming - unlike repair - doesn't do view building? And if it does as @vladzcloudius claims, is this a recent regression or for some reason we always did unnecessary view building in this case and somehow we never noticed it?

It should be easily reproducible.

**Comment by denesb** [comment]
I just checked the code again on how we store view build status.
The status of built-view is stored in `system_distributed.view_build_status` table, which has the following schema:
```
52  schema_ptr view_build_status() {                                                                                                                                                                                                                          
  1     static thread_local auto schema = [] {                                                                                                                                                                                                                
  2         auto id = generate_legacy_id(system_distributed_keyspace::NAME, system_distributed_keyspace::VIEW_BUILD_STATUS);                                                                                                                                  
  3         return schema_builder(system_distributed_keyspace::NAME, system_distributed_keyspace::VIEW_BUILD_STATUS, std::make_optional(id))                                                                                                                  
  4                 .with_column("keyspace_name", utf8_type, column_kind::partition_key)                                                                                                                                                                      
  5                 .with_column("view_name", utf8_type, column_kind::partition_key)                                                                                                                                                                          
  6                 .with_column("host_id", uuid_type, column_kind::clustering_key)                                                                                                                                                                           
  7                 .with_column("status", utf8_type)                                                                                                                                                                                                         
  8                 .with_version(system_keyspace::generate_schema_version(id))                                                                                                                                                                               
  9                 .build();                                                                                                                                                                                                                                 
 10     }();                                                                                                                                                                                                                                                  
 11     return schema;                                                                                                                                                                                                                                        
 12 }     
 ```
The system distributed keyspace itself uses SimpleStrategy, with RF=3.
All reads and writes to this table happen with CL=ONE.
This is the code determining whether view update generation is necessary:
```
2039 future<bool> check_view_build_ongoing(db::system_distributed_keyspace& sys_dist_ks, const sstring& ks_name, const sstring& cf_name) {                                                                                                                    
   1     return sys_dist_ks.view_status(ks_name, cf_name).then([] (std::unordered_map<utils::UUID, sstring>&& view_statuses) {                                                                                                                                
   2         return boost::algorithm::any_of(view_statuses | boost::adaptors::map_values, [] (const sstring& view_status) {                                                                                                                                   
   3             return view_status == "STARTED";                                                                                                                                                                                                             
   4         });                                                                                                                                                                                                                                              
   5     });                                                                                                                                                                                                                                                  
   6 }                                                                                                                                                                                                                                                        
   7                                                                                                                                                                                                                                                          
   8 future<bool> check_needs_view_update_path(db::system_distributed_keyspace& sys_dist_ks, const table& t, streaming::stream_reason reason) {                                                                                                               
   9     if (is_internal_keyspace(t.schema()->ks_name())) {                                                                                                                                                                                                   
  10         return make_ready_future<bool>(false);                                                                                                                                                                                                           
  11     }                                                                                                                                                                                                                                                    
  12     if (reason == streaming::stream_reason::repair && !t.views().empty()) {                                                                                                                                                                              
  13         return make_ready_future<bool>(true);                                                                                                                                                                                                            
  14     }                                                                                                                                                                                                                                                    
  15     return do_with(t.views(), [&sys_dist_ks] (auto& views) {                                                                                                                                                                                             
  16         return map_reduce(views,                                                                                                                                                                                                                         
  17                 [&sys_dist_ks] (const view_ptr& view) { return check_view_build_ongoing(sys_dist_ks, view->ks_name(), view->cf_name()); },                                                                                                               
  18                 false,                                                                                                                                                                                                                                   
  19                 std::logical_or<bool>());                                                                                                                                                                                                                
  20     });                                                                                                                                                                                                                                                  
  21 }      
  ```
Like I mentioned all reads and writes to the `view_build_status` happen with CL=ONE, so it is entirely possible that some replica observes the write with the status updated to `STARTED`, then misses the later write updating the status to `DONE`. If this replica is later queried for the status of the view, it will return `STARTED` and the inquiring node will decide that update generation is necessary.

Do we ever repair this keyspace during the reguler cluster maintenance?

**Comment by vladzcloudius** [comment]
Following @asias post here: https://github.com/scylladb/scylladb/issues/7711#issuecomment-1287444686

Asias, the problem is not whether MV is streamed or not - it's why a *base table* streamed data goes into `staging` and triggers a corresponding view update for every partition that is streamed.

The test you have performed doesn't trigger the interesting flow - you have created a local MV by using the same PK as the base table. 

The issue that we saw was observed with a view that has a partition key different from the base and under the condition of I/O slow enough not to be able to handle the incoming flow.

So, if you want to reproduce the problem that we saw you'd rather:
* artificially decrease the I/O capabilities by reducing values in io_properties.yaml
* Create a few MVs for the same base table with a PK different from the base (e.g. set it to be "Ci", where 'i' is 1,2,3,4,5 for each of those MVs).
* And when you run the test check what's going on in the `staging` directory of the base table.
* Use Scylla nodes with at least 16 vCPUs.

**Comment by vladzcloudius** [comment]
> Do we ever repair this keyspace during the reguler cluster maintenance?

`system_distributed` IS repaired but how does it help us, @denesb ? (no need to answer this rhetorical question)

If we care to read what we wrote we need consistencies of corresponding operations satisfy a well known formula: W + R > RF.
Using `QUORUM` or `EACH_QUORUM` for writes and `QUORUM`/`LOCAL_QUORUM` for reads would do the trick.

**Comment by denesb** [comment]
> `system_distributed` IS repaired but how does it help us, @denesb ? (no need to answer this rhetorical question)

This is very much not a rhetorical question and is relevant to all our current versions. It helps us because the as I explained above in detail, the code which determines whether streaming needs to generate view updates reads the system distributed keyspace with CL=ONE to determine this (after earlier writes with CL=ONE). So without a repair, it is very much possible for streaming to read a stale status and make the wrong decision that view updates have to be generated, moving all written sstables to the staging directory.
 
> If we care to read what we wrote we need consistencies of corresponding operations satisfy a well known formula: W + R > RF. Using `QUORUM` or `EACH_QUORUM` for writes and `QUORUM`/`LOCAL_QUORUM` for reads would do the trick.

I agree and we should change the code, unless the original author had a very good reason to use CL=ONE. In any case such a change is risky and is unlikely to be backported to 2021.1, so on that version we'll have the repair of the system distributed keyspace as a workaround before node ops.

**Comment by nyh** [comment]
@denesb is this a duplicate of #11905 (or the opposite ;-)), and fixed by your 84a69b6adb3df64657956404884c3880102e84b4 ?

**Comment by denesb** [comment]
> @denesb is this a duplicate of #11905 (or the opposite ;-)), and fixed by your [84a69b6](https://github.com/scylladb/scylladb/commit/84a69b6adb3df64657956404884c3880102e84b4) ?

Yes, I guess we could say it is a duplicate. Actually this issue is a super-set of #11905, the latter being the actual root cause why we did streaming when we shouldn't have.

**Comment by mykaul** [comment]
Closing as dup, thanks.
